Refactors FilteredQueryBuilder

MaineC opened the pull request:
Separates JSON parsing from Lucene query creation, adds support for streaming, hashCode and equals as well as unit tests.

Relates to #10217 

@cbuescher assigning to you for a first round of reviews. Not particularly happy with the createExpectedQuery implementation at the moment.

cbuescher commented on core/src/main/java/org/elasticsearch/index/query/FilteredQueryBuilder.java:
```diff
@@ -83,9 +158,7 @@ protected void doXContent(XContentBuilder builder, Params params) throws IOExcep
-        if (boost != -1) {
-            builder.field("boost", boost);
-        }
+        builder.field("boost", boost);
```
We changed the boost default to 1.0f in the other builders, should we do this here to for consistency? Can't add this comment on the top where the default is set.

cbuescher commented on core/src/main/java/org/elasticsearch/index/query/FilteredQueryBuilder.java:
```diff
@@ -72,6 +95,58 @@ public FilteredQueryBuilder queryName(String queryName) {
+            return queryBuilder.toQuery(parseContext);
+        }
+
+        Query query = queryBuilder.toQuery(parseContext);
```
I think we need additional null checks here. Before the refactoring we only needed to check lucene queries in parse() method for null. Now also their `toQuery()` can potentially return `null` (e.g. if this FilteredQueryBuilder would be nested in itself). Same goes for the filter.

cbuescher replied:
btw. I am currently thinking about how we can at least make the inner QueryBuilders we use always non-null, but will need to do this in separate PR and ask Adrien about it.

cbuescher commented on core/src/main/java/org/elasticsearch/index/query/FilteredQueryParser.java:
```diff
@@ -63,10 +60,9 @@ public Query parse(QueryParseContext parseContext) throws IOException, QueryPars
                 } else if ("filter".equals(currentFieldName)) {
-                    filterFound = true;
-                    filter = parseContext.parseInnerFilter();
+                    filter = parseContext.parseInnerQueryBuilder();
```
This needs to be another method (`parseInnerFilterToQueryBuilder`) which replaces `parseInnerFilter` and also takes care of switching the interal `isFilter` flag.

cbuescher commented on core/src/test/java/org/elasticsearch/index/query/FilteredQueryBuilderTest.java:
```diff
@@ -0,0 +1,78 @@
+    @SuppressWarnings("rawtypes")
+    @Override
+    protected FilteredQueryBuilder createTestQueryBuilder() {
+        QueryBuilder queryBuilder = RandomQueryBuilder.create(random());
```
So currently `RandomQueryBuilder.create(random())` doesn't return `null`, but I see you already plan for this in `testToQuery`. Maybe don't throw the switch in the random query generator just yet because I'm thinking about changes there (see comment above), but maybe randomly add `null` just here in createTestQueryBuilder so it won't blow up the other tests.

MaineC replied:
Thanks for this remark - currently in the process of fixing an issue I ran into after adjusting the test...

cbuescher commented on core/src/main/java/org/elasticsearch/index/query/FilteredQueryBuilder.java:
```diff
@@ -72,6 +95,58 @@ public FilteredQueryBuilder queryName(String queryName) {
+
+        BooleanQuery filteredQuery = Queries.filtered(queryBuilder.toQuery(parseContext), filter);
+        filteredQuery.setBoost(boost);
+        if (queryName != null) {
```
I just realize that there might be a bug in the existing code already. We only seem to add the query to the named queries if it's a BooleanQuery, the other cases return early. Not sure, but we might want to change that.

cbuescher commented:
I  went through it and it looks like there only a few minor changes to address. Left a few comments.

MaineC commented:
Thanks for your comments. Looking into the changes needed.

MaineC commented:
@cbuescher I believe I addressed your comments. Added call to AbstractQueryBuilder#doXContentInnerBuilder in doXContent - happy to remove if #11915 comes in before this one.

cbuescher commented on core/src/main/java/org/elasticsearch/index/query/FilteredQueryBuilder.java:
```diff
@@ -72,20 +96,64 @@ public FilteredQueryBuilder queryName(String queryName) {
+    }
+
+    @Override
+    public Query toQuery(QueryParseContext parseContext) throws QueryParsingException, IOException {
```
Can't find `validate()` here. I think we went for always having empty implementation in builder class, even if there's nothing to validate.

javanna replied:
it wouldn't be empty here at this point, it needs to validate the inner query and filter, see #11889

javanna replied:
this will need to become protected doToQuery as soon as you rebase

cbuescher commented:
Left one very minor comment, otherwise LGTM. Some of the code concerning null-checks for inner queries can be changed once #11915 is on feature branch. @javanna can you also have a look?

javanna commented on core/src/main/java/org/elasticsearch/common/io/stream/StreamOutput.java:
```diff
@@ -454,15 +454,28 @@ public void writeThrowable(Throwable throwable) throws IOException {
     /**
+     * Writes an optiona {@link NamedWriteable} to the current stream.
+     * */
+    public void writeOptionalNamedWriteable(@SuppressWarnings("rawtypes") NamedWriteable namedWriteable) throws IOException {
```
I don't see why the suppress warnings are needed here and in the previous method. Can you clarify that?

MaineC replied:
NamedWritable is a parameterised type so the compiler will complain when used w/o a parameter.

There's a few ways out: Specify the exact type of NamedWriteable that will be written in this location - clearly not an option at compile time here. Specify <?> instead of leaving the raw type - this would trickle down to any classes using this method, including e.g. the list of must clauses in BoolQueryBuilder and imply that all entries in that list need to be of the same type - not a good idea neither. The only way to get rid of the warning I saw was to disable with SuppressWarnings. (javac -X gave the hint on which warnings the JDK I use supports as part of the static analysis)

Checked the codebase for other uses of the annotation found it only in org/elasticsearch/search/aggregations/metrics/ScriptedMetricTests.java and org/elasticsearch/search/aggregations/pipeline/SiblingPipelineAggregator.java so I suppose we don't care too much about it. Happy to remove the annotations I added.

javanna replied:
alright, I do get unchecked warnings when it comes to generics, but not these ones, it is just the level of the warnings in the IDE. I agree we don't usually suppress these, I think we can remove.

javanna commented on core/src/main/java/org/elasticsearch/common/io/stream/StreamOutput.java:
```diff
@@ -454,15 +454,28 @@ public void writeThrowable(Throwable throwable) throws IOException {
      * Writes a list of {@link NamedWriteable} to the current stream, by first writing its size and then iterating over the objects
      * in the list
      */
+    @SuppressWarnings("rawtypes") 
```
here too, I don't get why

javanna commented on core/src/main/java/org/elasticsearch/index/query/FilteredQueryBuilder.java:
```diff
@@ -19,17 +19,26 @@
  * @deprecated Use {@link BoolQueryBuilder} instead.
  */
 @Deprecated
+@SuppressWarnings("rawtypes")
```
suppress warnings here? why is it needed?

javanna commented on core/src/main/java/org/elasticsearch/index/query/FilteredQueryBuilder.java:
```diff
@@ -72,20 +96,64 @@ public FilteredQueryBuilder queryName(String queryName) {
+            result = query;
+        } else if (Queries.isConstantMatchAllQuery(filter)) {
+            // this is an instance of match all filter, just execute the query
+            result = query;
```
we lose the boost here and in the case before? maybe the setBoost should be at the end of the if?

MaineC replied:
We lost the boost before as well - I wasn't sure this was intentional or a bug, fixing.

javanna replied:
no worries as soon as you rebase you won't need to take care of  boost and _name, it's done automatically.

javanna commented on core/src/main/java/org/elasticsearch/index/query/FilteredQueryBuilder.java:
```diff
@@ -72,20 +96,64 @@ public FilteredQueryBuilder queryName(String queryName) {
+    protected void doXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject(NAME);
+        super.doXContentInnerBuilder(builder, "query", queryBuilder, params);
+        super.doXContentInnerBuilder(builder, "filter", filterBuilder, params);
```
the super prefix is not needed here right?

javanna commented on core/src/main/java/org/elasticsearch/index/query/FilteredQueryParser.java:
```diff
@@ -43,13 +40,13 @@ public FilteredQueryParser() {
-        Query query = Queries.newMatchAllQuery();
-        Query filter = null;
-        boolean filterFound = false;
+        QueryBuilder<MatchAllQueryBuilder> query = new MatchAllQueryBuilder();
```
the use of generics here is fishy. You wanted to assign a default value to the query I think? But then the query builder that gets parsed later on does not necessarilyhave MatchAllQueryBuilder type?

javanna commented on core/src/test/java/org/elasticsearch/index/query/BaseQueryTestCase.java:
```diff
@@ -19,6 +19,8 @@
 
 package org.elasticsearch.index.query;
 
+import com.carrotsearch.randomizedtesting.annotations.Repeat;
```
this is a leftover I think

javanna commented on core/src/test/java/org/elasticsearch/index/query/BaseQueryTestCase.java:
```diff
@@ -198,17 +200,17 @@ public void afterTest() {
     @Test
     public void testFromXContent() throws IOException {
-        QB testQuery = createTestQueryBuilder();
+        QB expectedQuery = createTestQueryBuilder();
```
not sure about this name change to be honest. This query is the one that we use for testing, we start with it, we convert it to string, we reparse it and we compare the result with newQuery, which is actually the expected query.

javanna commented on core/src/test/java/org/elasticsearch/index/query/BaseQueryTestCase.java:
```diff
@@ -223,9 +225,13 @@ public void testToQuery() throws IOException {
-        assertThat(actualQuery, instanceOf(expectedQuery.getClass()));
-        assertThat(actualQuery, equalTo(expectedQuery));
-        assertLuceneQuery(testQuery, actualQuery, context);
+        if (expectedQuery == null) {
```
mmm why would createTestQueryBuilder return null?

MaineC replied:
In the original FilteredQueryParser.parse(...) if both nested queries were parsed to null we returned null. This will change with #11915

cbuescher replied:
I also need this already in #11915 when I add possibility of tests with inner query builders that return null (EmptyQueryBuilder).

javanna replied:
gotcha

javanna commented on core/src/test/java/org/elasticsearch/index/query/FilteredQueryBuilderTest.java:
```diff
@@ -0,0 +1,85 @@
+
+        BooleanQuery filteredQuery = Queries.filtered(qb.query().toQuery(context), filter);
+        filteredQuery.setBoost(qb.boost());
+        return filteredQuery;
```
don't worry too much about this method, we wil get rid of it soon. I am preparing the field for this change :)

javanna commented on core/src/main/java/org/elasticsearch/index/query/FilteredQueryBuilder.java:
```diff
@@ -72,20 +96,64 @@ public FilteredQueryBuilder queryName(String queryName) {
+    @Override
+    public Query toQuery(QueryParseContext parseContext) throws QueryParsingException, IOException {
+
+        Query query = (queryBuilder == null) ? null : queryBuilder.toQuery(parseContext);
```
I think the null handling should need to be aligned with #11915 somehow, not sure which one of the two PRs should get in first. As for the behaviour, is it correct that none of the inner query is mandatory at the moment (before the refactoring) and that doesn't change here? That means match_all query right?

MaineC replied:
I believe I didn't change the logic (other than making sure that any generated query gets added to named queries, and after this review round gets the boost applied) - would be great to get one final check from @jpountz though as I changed the way the final Lucene query is generated quite a bit though.

> As for the behaviour, is it correct that none of the inner query is mandatory at the moment 
> (before the refactoring) and that doesn't change here? That means match_all query right?

Nearly right: Originally in the parser we started with a match_all query. We than parsed both, the filter and query part. If either returned a null value, we ended up returning null as the final query.

The match_all somewhat-default wasn't (and isn't) built into the FilteredQueryBuilder though: Users always had to got through the constructor that required to specify both, query and filter but allowed both to be null. 

Now trying to get the knot out of my head.

MaineC commented:
Left some explanations and updated the PR.

javanna commented on core/src/main/java/org/elasticsearch/index/query/FilteredQueryBuilder.java:
```diff
@@ -30,6 +36,8 @@
 @Deprecated
 public class FilteredQueryBuilder extends AbstractQueryBuilder<FilteredQueryBuilder> implements BoostableQueryBuilder<FilteredQueryBuilder> {
+    /** Default for boost of this query. */
+    public static final float DEFAULT_BOOST = 1.0f;
```
you can now use the default coming from AbstractQueryBuilder instead

javanna commented on core/src/main/java/org/elasticsearch/index/query/FilteredQueryBuilder.java:
```diff
@@ -72,20 +95,72 @@ public FilteredQueryBuilder queryName(String queryName) {
+    public String queryName() {
+        return queryName;
+    }
+
```
rebase and these getters wont be needed anymore ;)

javanna commented on core/src/main/java/org/elasticsearch/index/query/FilteredQueryBuilder.java:
```diff
@@ -72,20 +95,72 @@ public FilteredQueryBuilder queryName(String queryName) {
+        return Objects.equals(queryBuilder, other.queryBuilder) &&
+                Objects.equals(filterBuilder, other.filterBuilder) &&
+                Objects.equals(boost, other.boost) &&
+                Objects.equals(queryName, other.queryName);
```
once you rebase you can take out boost and queryName here, they are already taken care of in the base class

javanna commented on core/src/main/java/org/elasticsearch/index/query/FilteredQueryBuilder.java:
```diff
@@ -72,20 +95,72 @@ public FilteredQueryBuilder queryName(String queryName) {
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(queryBuilder, filterBuilder, boost, queryName);
```
once you rebase you need to implement doHashCode instead, and can take out boost and queryName here, they are already taken care of in the base class,

javanna commented on core/src/main/java/org/elasticsearch/index/query/FilteredQueryBuilder.java:
```diff
@@ -72,20 +95,72 @@ public FilteredQueryBuilder queryName(String queryName) {
+        builder.field("boost", boost);
         if (queryName != null) {
             builder.field("_name", queryName);
         }
```
you can use the util method printBoostAndQueryName here now (once you rebased)

javanna commented on core/src/main/java/org/elasticsearch/index/query/FilteredQueryBuilder.java:
```diff
@@ -96,4 +171,22 @@ protected void doXContent(XContentBuilder builder, Params params) throws IOExcep
+        out.writeOptionalNamedWriteable(queryBuilder);
+        out.writeOptionalNamedWriteable(filterBuilder);
+        out.writeFloat(boost);
+        out.writeOptionalString(queryName);
```
take out boost and queryname once you rebased

javanna commented on core/src/main/java/org/elasticsearch/index/query/FilteredQueryBuilder.java:
```diff
@@ -96,4 +171,22 @@ protected void doXContent(XContentBuilder builder, Params params) throws IOExcep
+        QueryBuilder filter = in.readOptionalNamedWriteable();
+        FilteredQueryBuilder qb = new FilteredQueryBuilder(query, filter);
+        qb.boost = in.readFloat();
+        qb.queryName = in.readOptionalString();
```
take out boost and queryname once you rebased

javanna commented:
I did another round, looks good but I would need to have another look after you rebased I think, #11974 changed things quite a bit (for the better!)

MaineC commented:
Did another round after re-basing.

@jpountz if you have a few free cycles it would be good if you could take a final look at the way the Lucene query is generated as well as I restructured it quite a bit.

javanna commented on core/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java:
```diff
@@ -568,6 +568,18 @@ public Object readGenericValue() throws IOException {
+     * */
+    public <C> C readOptionalNamedWriteable() throws IOException {
+        boolean readable = readBoolean();
+        if (readable) {
```
super minor thing, but here we could do if (readBoolean()) directly  without the readable variable

javanna commented on core/src/main/java/org/elasticsearch/index/query/FilteredQueryParser.java:
```diff
@@ -83,44 +78,15 @@ public Query parse(QueryParseContext parseContext) throws IOException, QueryPars
-        // parsed internally, but returned null during parsing...
-        if (query == null) {
-            return null;
-        }
```
this null check seems useful, if no query was found, return null meaning ignore this filtered query as a whole. weird though.

javanna replied:
ignore this comment for now :)

javanna commented on core/src/main/java/org/elasticsearch/index/query/FilteredQueryParser.java:
```diff
@@ -83,44 +78,15 @@ public Query parse(QueryParseContext parseContext) throws IOException, QueryPars
-        if (queryName != null) {
-            parseContext.addNamedQuery(queryName, filteredQuery);
-        }
-        return filteredQuery;
```
what doesn't convince me about removing all these checks is that there is a difference between returning null and returning match_all I think, which we had before. null means ignore (no-op), while match_all is a query that matches all documents. Also this is going to change again with #11915. I would rather wait, get #11915 in and then refactor this one once and forever :)

MaineC replied:
Sure - waiting for #11915 makes sense here.

About null vs. match_all: Before we would return match_all in the parser if the query part of the filter was missing in the json definition. We would return null if the query part was present but parsed to null. 

Going through the REST API this is still true as the query part gets initialised with a MatchAllQueryBuilder.

Maybe the right thing to do to carry that semantic over to the Java API would be to also initialize query to MatchAllQueryBuilder there but let the user set it to null if they want to.

javanna commented on core/src/test/java/org/elasticsearch/index/query/FilteredQueryBuilderTest.java:
```diff
@@ -0,0 +1,84 @@
+        }
+        if (randomBoolean()) {
+            query.queryName(randomUnicodeOfLengthBetween(3, 15));
+        }
```
you can remove randomization of boost and queryName

javanna commented on core/src/test/java/org/elasticsearch/index/query/FilteredQueryBuilderTest.java:
```diff
@@ -0,0 +1,84 @@
+        }
+
+        BooleanQuery filteredQuery = Queries.filtered(qb.query().toQuery(context), filter);
+        filteredQuery.setBoost(qb.boost());
```
remove the setBoost

javanna commented on core/src/test/java/org/elasticsearch/index/query/FilteredQueryBuilderTest.java:
```diff
@@ -0,0 +1,84 @@
+        // if its a match_all query, use constant_score
+        if (Queries.isConstantMatchAllQuery(query)) {
+            Query q = new ConstantScoreQuery(filter);
+            q.setBoost(qb.boost());
```
remove the set boost

javanna commented on core/src/test/java/org/elasticsearch/index/query/BaseQueryTestCase.java:
```diff
@@ -218,8 +218,8 @@ public void testFromXContent() throws IOException {
         assertNotSame(newQuery, testQuery);
-        assertEquals("Queries should be equal", testQuery, newQuery);
-        assertEquals("Queries should have equal hashcodes", testQuery.hashCode(), newQuery.hashCode());
+        assertEquals("Queries should be equal: " + testQuery + " vs. " + newQuery, testQuery, newQuery);
```
I think I removed this testQuery vs. newQuery. when it fails I see the same message twice:
testQuery vs. newQuery and then
expected: testQuery, actual: newQuery

I don't see the added value of this, maybe it's my IDE or something?

jpountz commented:
> @jpountz if you have a few free cycles it would be good if you could take a final look at the way the Lucene query is generated as well as I restructured it quite a bit.

This looks good to me!

MaineC commented:
@jpountz Thanks for the feedback.

javanna commented on core/src/main/java/org/elasticsearch/index/query/FilteredQueryBuilder.java:
```diff
@@ -40,6 +46,15 @@
+        this.queryBuilder = new MatchAllQueryBuilder();
+        this.filterBuilder = filterBuilder;
+    }
+
```
I think we have to guard against null values coming from the java api in all the constructors

cbuescher replied:
+1 for doing this in the old 2-arg constructor and forwarding to it here using the match_all default.

MaineC replied:
Happy to do this - should this be mentioned in the migration_query_refactoring doc as before null was allowed here?

javanna replied:
not sure, I don't see it as a breaking change cause it just makes sure that a user cannot do stupid things... @cbuescher what do you think?

cbuescher replied:
I just looked at the original parser code and the docs again and saw that it explicitly allows leaving out the `filter` section:
```
// we allow for null filter, so it makes compositions on the client side to be simpler
```
The JSON side of this query seems to be currently seems super lenient (basically both filter/query can be null), all null-values map to some defaults. If we restrict this on the JAVA-API, we should maybe be explicit about it in this case, saying which nested queries to use to mimic the JSON behaviour when leaving the clauses there.
I'm on the fence about adding null-checks in constructors after reading the original parser code, maybe we should also allow it in the Java-API but internally always set the defaults to avoid null-checks all over the place. Sorry for this back & forth.

javanna replied:
but the parser will never set null with out refactoring right? So we could have defaults in the builder and convert set null to set default value like we did in other place? Just looking for a way to stay away from null values

MaineC replied:
I think the question here boils down to how much of the optimization/default handling that was previously done in the parser should be moved to the builder and thus made available to users of the Java API.

If we want Java API and REST API to be similar so users can directly map from what they are doing with e.g. Sense to Java API calls I believe we should be lenient wrt. accepting null values in the Builder.

If we'd rather avoid handling null values in the Builder at all the resulting Java API will be more strict than what we accept through REST.

In the latest update to the PR I tried to find a middle ground - accepting null values but mapping them to EmptyQueryBuilder where possible.

javanna commented on core/src/main/java/org/elasticsearch/index/query/FilteredQueryBuilder.java:
```diff
@@ -40,6 +46,15 @@
+     * @param filterBuilder The filter to apply on the query (Can be null)
+     * */
+    public FilteredQueryBuilder(@Nullable QueryBuilder filterBuilder) {
+        this.queryBuilder = new MatchAllQueryBuilder();
```
shall we use MatchAllQueryBuilder as a default value in general, outside of this constructor? If not set, it will be match all? When set, we check that it's not set to null in setter.

cbuescher replied:
+1 on introducing a constant somewhere and reuse it in the parser as well.

MaineC replied:
Couldn't introduce a constant: MatchAllQueryBuilder has a setBoost/setQueryName, FilteredQueryBuilder in turn has a getQuery() so users could get access to the internal default query and change it's properties.

Went for a generateDefaultQuery() static method instead.

javanna commented on core/src/main/java/org/elasticsearch/index/query/FilteredQueryBuilder.java:
```diff
@@ -50,17 +65,68 @@ public FilteredQueryBuilder(@Nullable QueryBuilder queryBuilder, @Nullable Query
+        Query filter = filterBuilder.toQuery(parseContext);
+
+        if (query == null) {
+            return null;
```
thinking out loud here, this is tricky: previously we could have null in case the query is there but parsed as null, which we now have EmptyQueryBuilder for. EmptyQueryBuider#toQuery returns null, so this if seems to be right here. that said I find it confusing, cause we end up ignoring query and filter in this case. Why?

cbuescher replied:
Also think this is right here. In JSON you get null here for
```
{
  "query": {
    "filtered": { 
      "query": { },
      "filter": {
        "range": { "created": { "gte": "now - 1d / d" }}
      }
    }
  }
}
```
 Which means the user specified something as query but its the empty set. In this case its not clear what to filter, but also not save to return any of Match_All/No_Docs query, since that would mean something different depending on any (potential) enclosing query, e.g. when this is used in "must" or in "must_not".

javanna replied:
agreed my doubt is not about the refactoring, more about the general behaviour, because the inner query is empty we ignore the whole filtered query, despite a proper filter was there. anyway I would just double check that this is the same behaviour as before and add a clear comment on when and why this happens :)

MaineC replied:
It's the exact same behaviour as here:

https://github.com/elastic/elasticsearch/blob/master/core/src/main/java/org/elasticsearch/index/query/FilteredQueryParser.java#L90

I don't find that pretty but tried to keep the refactored code close to the original.

javanna replied:
alright that's what I thought too, sounds good

javanna commented on core/src/main/java/org/elasticsearch/index/query/FilteredQueryParser.java:
```diff
@@ -44,12 +41,11 @@ public FilteredQueryParser() {
-        Query query = Queries.newMatchAllQuery();
-        Query filter = null;
-        boolean filterFound = false;
+        QueryBuilder query = new MatchAllQueryBuilder();
```
if we end up using the match_all query builder twice as a default, maybe we should have a constant for this default in the builder?

javanna commented on core/src/test/java/org/elasticsearch/index/query/FilteredQueryBuilderTest.java:
```diff
@@ -0,0 +1,62 @@
+        result.setBoost(qb.boost());
+        return result;
+    }
+
```
we are missing a test for validate method here

javanna commented:
left a few comments, @cbuescher please have a look too :)

cbuescher commented:
Also went through the changes once again, would prefer null-checks in constructors as well, left some more remarks.

MaineC commented:
Went through your comments and updated the PR (noticed only after pushing that one of the questions I had left was answered by both of you in the mean time - you are too fast answering questions).

cbuescher commented:
@MaineC left a few comments, sorry I didn't realize I should not to that on the individual commits but rather on the overall change. Will do so next time.

javanna commented on .settings/org.eclipse.core.resources.prefs:
```diff
@@ -1,6 +1,3 @@
 eclipse.preferences.version=1
-encoding//src/main/java=UTF-8
-encoding//src/main/resources=UTF-8
-encoding//src/test/resources=UTF-8
```
what happened here?

MaineC replied:
Meh.

javanna commented on core/src/main/java/org/elasticsearch/index/query/FilteredQueryBuilder.java:
```diff
@@ -31,36 +36,113 @@
+        if (query == null) {
+            // Most likely this query was parsed from the JSON query DSL - it parsed to null so we ignore
+            // the whole filtered query as there is nothing to filter on. See FilteredQueryParser for an example.
+            return null;
```
I think we have to clarify this comment, it didn't parse to null, that never happens anymore (othersie you would have NPE at line 101). This became a null lucene query after calling toQuery.

javanna commented on core/src/main/java/org/elasticsearch/index/query/FilteredQueryBuilder.java:
```diff
@@ -31,36 +36,113 @@
+        }
+        if (filterBuilder == null) {
+            validationException = addValidationError("Filter in FilteredQueryBuilder cannot be null.", validationException);
+        }
```
I think filter and query can never be null here? not sure whether we should validate this here.

cbuescher replied:
I was having the same question in another query I was looking at. I think we made sure in constructors that the two builders are never null, but I was also wondering if if doesn't hurt having extra checks, in case e.g. some later change makes it possible to sneak in null query builders somehow. On the other hand, test should cover this. Sorry, undecided here, that's just what I had in mind in similar situation.

javanna replied:
right I think tests are made for that. I wouldn't want to have checks for something that we handle earlier and that should never happen here, that is my personal opinion though :)

javanna commented on core/src/main/java/org/elasticsearch/index/query/FilteredQueryBuilder.java:
```diff
@@ -31,36 +36,113 @@
+     * default queryBuilder if none is supplied by the user. Feel free to
+     * set queryName and boost on that that instance - it's always a new one.
+     * */
+    public static QueryBuilder generateDefaultQuery() {
```
I think this can be a constant rather than a method?

javanna replied:
a constant could cause problems though if the constant gets misused (e.g. changing its boost etc.), method is ok sorry

javanna commented on core/src/main/java/org/elasticsearch/index/query/FilteredQueryParser.java:
```diff
@@ -44,12 +41,11 @@ public FilteredQueryParser() {
-        Query query = Queries.newMatchAllQuery();
-        Query filter = null;
-        boolean filterFound = false;
+        QueryBuilder query = FilteredQueryBuilder.generateDefaultQuery();
```
maybe we could pass in null to the builder since the default is already handled there, that way the odd generateDefaultQuery could become private too :)

javanna commented on core/src/main/java/org/elasticsearch/index/query/FilteredQueryParser.java:
```diff
@@ -83,44 +78,35 @@ public Query parse(QueryParseContext parseContext) throws IOException, QueryPars
+             *   }
+             * }
+             * */
+            return EmptyQueryBuilder.PROTOTYPE;
```
this can be null only if the query element wasn't there at all I think, if it was there but empty it would become an EmptyQueryBuilder during `parseContext#parseInnerQueryBuilder`. I will leave it to @cbuescher though to decide whether what we do here is correct :) I am not sure

cbuescher replied:
Agree with this, I already left a comment about this further up the timeline (https://github.com/MaineC/elasticsearch/commit/fd127841d6c907980b5b0328422347e50e855ee7#commitcomment-12034514)

javanna commented on core/src/test/java/org/elasticsearch/index/query/BaseQueryTestCase.java:
```diff
@@ -218,8 +218,8 @@ public void testFromXContent() throws IOException {
-        assertEquals("Queries should be equal", testQuery, newQuery);
-        assertEquals("Queries should have equal hashcodes", testQuery.hashCode(), newQuery.hashCode());
+        assertEquals(testQuery, newQuery);
+        assertEquals("Queries should have equal hashcodes: " + testQuery + " vs. " + newQuery, testQuery.hashCode(), newQuery.hashCode());
```
I left a previous comment around this, I had removed this text cause for me it causes duplication of the error message. Do you see something different? Junit already prints out "expect: .... actual: .... " with this addition it would print the same stuff out twice, at least that is what I used to see

MaineC replied:
Without the extra message I only see the raw hashCode numbers that are different, but not the actual QueryBuilder object as seen through it's toString method. Can remove if that's not helpful though.

javanna replied:
right that makes sense when we test hashcode, sorry I missed that

javanna commented on core/src/test/java/org/elasticsearch/index/query/FilteredQueryBuilderTest.java:
```diff
@@ -0,0 +1,95 @@
+
+
+        qb = new FilteredQueryBuilder(invalid, valid);
+        assertNotNull(qb.validate());
```
count the expected errors too like we do in other tests? also we never do (invalid, invalid). I think randomizing things may improve this test and coverage too, like we do in other tests.

MaineC replied:
Makes sense.

javanna commented on core/src/main/java/org/elasticsearch/index/query/FilteredQueryParser.java:
```diff
@@ -83,44 +78,35 @@ public Query parse(QueryParseContext parseContext) throws IOException, QueryPars
-        if (Queries.isConstantMatchAllQuery(filter)) {
-            // this is an instance of match all filter, just execute the query
-            return query;
+            filter = EmptyQueryBuilder.PROTOTYPE;
```
I am wondering whether we should just use null here, which would become EmptyQueryBuilder anyway in the builder

javanna commented:
I did another round of review, left some comments. I'd need @cbuescher to have another look too though  :)

cbuescher commented on core/src/test/java/org/elasticsearch/index/query/FilteredQueryBuilderTest.java:
```diff
@@ -0,0 +1,95 @@
+        } else {
+            result = Queries.filtered(qb.query().toQuery(context), filter);
+        }
+        result.setBoost(qb.boost());
```
Boost will be set in BaseQueryTestCase. No harm here, but doppelt gemoppelt (I want to try to gradually introduce this expression into the english vocabulary like Kindergarten and Schadenfreude) ;-)

javanna replied:
:) good catch

cbuescher commented:
Also did another round of reviews and left some comments, mostly minor stuff. Agree that the query null-check in the parser doesn't need to be there, happy to clear this up more if there are questions remaining.

MaineC commented:
Ready for another look.

javanna commented on core/src/main/java/org/elasticsearch/index/query/FilteredQueryBuilder.java:
```diff
@@ -31,36 +36,107 @@
+            result = new ConstantScoreQuery(filter);
+        } else {
+            // use a BooleanQuery
+            result = Queries.filtered(query, filter);
```
minor thing but how about getting rid of the result variable and just return the query from the if?

javanna commented on core/src/main/java/org/elasticsearch/index/query/FilteredQueryBuilder.java:
```diff
@@ -31,36 +36,107 @@
     /**
+     * Returns a {@link MatchAllQueryBuilder} instance that will be used as
+     * default queryBuilder if none is supplied by the user. Feel free to
+     * set queryName and boost on that that instance - it's always a new one.
```
that that :)

javanna commented:
LGTM besides the two minor comments I left
